### PR TITLE
Fix absolute links https://docs.microsoft.com/dotnet links to relative links

### PR DESCRIPTION
**Pull request by msschl**
## Summary

Changed absolute https://docs.microsoft.com/dotnet links to relative links.

Contributes to #8364

edit by @mairaw: updated keyword and issue number

### CODE REVIEW

File: styleguide/template.md [hunk 1]
```diff
@@ -454,7 +454,7 @@ A custom style is available for lists. You can render lists with green check mar
 > * How to add a class and an XmlSerializer
 > * How to build and run the application
 
-You can see an example of checked lists in action in the [.NET Core docs](https://docs.microsoft.com/dotnet/core/additional-tools/xml-serializer-generator).
+You can see an example of checked lists in action in the [.NET Core docs](~/docs/core/additional-tools/xml-serializer-generator.md).
```
Comment (mairaw) [inline styleguide/template.md]:
```suggestion
You can see an example of checked lists in action in the [.NET Core docs](https://docs.microsoft.com/dotnet/core/additional-tools/xml-serializer-generator).
```

File: styleguide/template.md [hunk 2]
```diff
@@ -477,4 +477,4 @@ You can see an example of selectors in action at the [Azure docs](https://docs.m
 >[Previous](../docs/csharp/expression-trees-interpreting.md)
 >[Next](../docs/csharp/expression-trees-translating.md)
 
-You can see an example of step-by-steps in action at the [C# Guide](https://docs.microsoft.com/dotnet/csharp/tour-of-csharp/program-structure).
+You can see an example of step-by-steps in action at the [C# Guide](~/docs/csharp/tour-of-csharp/program-structure.md).
```
Comment (mairaw) [inline styleguide/template.md]:
```suggestion
You can see an example of step-by-steps in action at the [C# Guide](https://docs.microsoft.com/dotnet/csharp/tour-of-csharp/program-structure).
```

**Review by mairaw — CHANGES_REQUESTED** [review]
Changes look great @msschl but the style guide page should use the absolute path since that page is only used here on GitHub and not part of docs, so the ~ notation won't work.

**Comment by msschl** [comment]
@mairaw reverted links to absolute links in `styleguide/template.md`.

File: docs/standard/microservices-architecture/microservice-ddd-cqrs-patterns/seedwork-domain-model-base-classes-interfaces.md [hunk 3]
```diff
@@ -19,7 +19,7 @@ This is the type of copy and paste reuse that many developers share between proj
 
 ## The custom Entity base class
 
-The following code is an example of an Entity base class where you can place code that can be used the same way by any domain entity, such as the entity ID, [equality operators](https://docs.microsoft.com/dotnet/csharp/language-reference/operators/equality-comparison-operator), a domain event list per entity, etc.
+The following code is an example of an Entity base class where you can place code that can be used the same way by any domain entity, such as the entity ID, [equality operators](~/docs/csharp/language-reference/operators/equality-comparison-operator.md), a domain event list per entity, etc.
```
Comment (NextTurn) [inline docs/standard/microservices-architecture/microservice-ddd-cqrs-patterns/seedwork-domain-model-base-classes-interfaces.md]:
```suggestion
The following code is an example of an Entity base class where you can place code that can be used the same way by any domain entity, such as the entity ID, [equality operators](../../../../csharp/language-reference/operators/equality-comparison-operator.md), a domain event list per entity, etc.
```

File: docs/standard/microservices-architecture/multi-container-microservice-net-applications/test-aspnet-core-services-web-apps.md [hunk 4]
```diff
@@ -104,7 +104,7 @@ public class PrimeWebDefaultRequestShould
     [*https://docs.microsoft.com/aspnet/core/test/integration-tests*](https://docs.microsoft.com/aspnet/core/test/integration-tests)
 
 -   **Unit testing in .NET Core using dotnet test** <br/>
-    [*https://docs.microsoft.com/dotnet/core/testing/unit-testing-with-dotnet-test*](https://docs.microsoft.com/dotnet/core/testing/unit-testing-with-dotnet-test)
+    [*https://docs.microsoft.com/dotnet/core/testing/unit-testing-with-dotnet-test*](~/docs/core/testing/unit-testing-with-dotnet-test.md)
```
Comment (NextTurn) [inline docs/standard/microservices-architecture/multi-container-microservice-net-applications/test-aspnet-core-services-web-apps.md]:
```suggestion
    [*https://docs.microsoft.com/dotnet/core/testing/unit-testing-with-dotnet-test*](../../../../core/testing/unit-testing-with-dotnet-test.md)
```

File: includes/migration-guide/retargeting/winforms/winforms-domain-upbutton-downbutton-actions-are-sync-now.md [hunk 5]
```diff
@@ -3,7 +3,7 @@
...
+|Suggestion|In order for an application to benefit from these changes, it must run on the .NET Framework 4.7.2 or later. The application can benefit from these changes in either of the following ways:<ul><li>It is recompiled to target the .NET Framework 4.7.2. This change is enabled by default on Windows Forms applications that target the .NET Framework 4.7.2 or later.</li><li>It opts out of the legacy scrolling behavior by adding the following [AppContext Switch](~/docs/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element.md) to the <code>&lt;runtime&gt;</code> section of the app config file and setting it to <code>false</code>, as the following example shows.</li></ul><pre><code class="lang-xml">&lt;runtime&gt;&#13;&#10;&lt;AppContextSwitchOverrides value=&quot;Switch.System.Windows.Forms.DomainUpDown.UseLegacyScrolling=false&quot;/&gt;&#13;&#10;&lt;/runtime&gt;&#13;&#10;</code></pre>|
```
Comment (NextTurn) [inline includes/migration-guide/retargeting/winforms/winforms-domain-upbutton-downbutton-actions-are-sync-now.md]:
```suggestion
|Suggestion|In order for an application to benefit from these changes, it must run on the .NET Framework 4.7.2 or later. The application can benefit from these changes in either of the following ways:<ul><li>It is recompiled to target the .NET Framework 4.7.2. This change is enabled by default on Windows Forms applications that target the .NET Framework 4.7.2 or later.</li><li>It opts out of the legacy scrolling behavior by adding the following [AppContext Switch](../../../../../docs/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element.md) to the <code>&lt;runtime&gt;</code> section of the app config file and setting it to <code>false</code>, as the following example shows.</li></ul><pre><code class="lang-xml">&lt;runtime&gt;&#13;&#10;&lt;AppContextSwitchOverrides value=&quot;Switch.System.Windows.Forms.DomainUpDown.UseLegacyScrolling=false&quot;/&gt;&#13;&#10;&lt;/runtime&gt;&#13;&#10;</code></pre>|
```

File: includes/migration-guide/retargeting/winforms/privatefontcollectionaddfontfile-method-releases-font-resources.md [hunk 6]
```diff
@@ -3,7 +3,7 @@
...
...In order for an application to benefit from these changes, it must run on the .NET Framework 4.7.2 or later. The application can benefit from these changes in either of the following ways:<ul><li>It is recompiled to target the .NET Framework 4.7.2. This change is enabled by default on Windows Forms applications that target the .NET Framework 4.7.2 or later.</li><li>It targets the .NET Framework 4.7.1 or an earlier version and opts out of the legacy accessibility behaviors by adding the following [AppContext Switch](~/docs/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element.md) to the <code>&lt;runtime&gt;</code> section of the app.config file and setting it to <code>false</code>, as the following example shows.</li></ul><pre><code class="lang-xml">&lt;runtime&gt;&#13;&#10;&lt;AppContextSwitchOverrides value=&quot;Switch.System.Drawing.Text.DoNotRemoveGdiFontsResourcesFromFontCollection=false&quot;/&gt;&#13;&#10;&lt;/runtime&gt;&#13;&#10;</code></pre>Applications that target the .NET Framework 4.7.2 or later, and want to preserve the legacy behavior can opt in to not release font resources by explicitly setting this AppContext switch to <code>true</code>.|
```
Comment (NextTurn) [inline includes/migration-guide/retargeting/winforms/privatefontcollectionaddfontfile-method-releases-font-resources.md]:
```suggestion
|Suggestion|<strong>How to opt in or out of these changes</strong>In order for an application to benefit from these changes, it must run on the .NET Framework 4.7.2 or later. The application can benefit from these changes in either of the following ways:<ul><li>It is recompiled to target the .NET Framework 4.7.2. This change is enabled by default on Windows Forms applications that target the .NET Framework 4.7.2 or later.</li><li>It targets the .NET Framework 4.7.1 or an earlier version and opts out of the legacy accessibility behaviors by adding the following [AppContext Switch](../../../../../docs/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element.md) to the <code>&lt;runtime&gt;</code> section of the app.config file and setting it to <code>false</code>, as the following example shows.</li></ul><pre><code class="lang-xml">&lt;runtime&gt;&#13;&#10;&lt;AppContextSwitchOverrides value=&quot;Switch.System.Drawing.Text.DoNotRemoveGdiFontsResourcesFromFontCollection=false&quot;/&gt;&#13;&#10;&lt;/runtime&gt;&#13;&#10;</code></pre>Applications that target the .NET Framework 4.7.2 or later, and want to preserve the legacy behavior can opt in to not release font resources by explicitly setting this AppContext switch to <code>true</code>.|
```

File: includes/migration-guide/retargeting/winforms/contextmenustripsourcecontrol-property-contains-valid-control-case-nested.md [hunk 7]
```diff
@@ -3,7 +3,7 @@
...
...g>In order for an application to benefit from these changes, it must run on the .NET Framework 4.7.2 or later. The application can benefit from these changes in either of the following ways:<ul><li>It targets the .NET Framework 4.7.2. This change is enabled by default on Windows Forms applications that target the .NET Framework 4.7.2 or later.</li><li>It targets the .NET Framework 4.7.1 or an earlier version and opts out of the legacy accessibility behaviors by adding the following [AppContext Switch](~/docs/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element.md) to the <code>&lt;runtime&gt;</code> section of the app.config file and setting it to <code>false</code>, as the following example shows.</li></ul><pre><code class="lang-xml">&lt;runtime&gt;&#13;&#10;&lt;AppContextSwitchOverrides value=&quot;Switch.System.Windows.Forms.UseLegacyContextMenuStripSourceControlValue=false&quot;/&gt;&#13;&#10;&lt;/runtime&gt;&#13;&#10;</code></pre>Applications that target the .NET Framework 4.7.2 or later, and want to preserve the legacy behavior can opt in to the use of the legacy source control value by explicitly setting this AppContext switch to <code>true</code>.|
```
Comment (NextTurn) [inline includes/migration-guide/retargeting/winforms/contextmenustripsourcecontrol-property-contains-valid-control-case-nested.md]:
```suggestion
|Suggestion|<strong>How to opt in or out of these changes</strong>In order for an application to benefit from these changes, it must run on the .NET Framework 4.7.2 or later. The application can benefit from these changes in either of the following ways:<ul><li>It targets the .NET Framework 4.7.2. This change is enabled by default on Windows Forms applications that target the .NET Framework 4.7.2 or later.</li><li>It targets the .NET Framework 4.7.1 or an earlier version and opts out of the legacy accessibility behaviors by adding the following [AppContext Switch](../../../../../docs/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element.md) to the <code>&lt;runtime&gt;</code> section of the app.config file and setting it to <code>false</code>, as the following example shows.</li></ul><pre><code class="lang-xml">&lt;runtime&gt;&#13;&#10;&lt;AppContextSwitchOverrides value=&quot;Switch.System.Windows.Forms.UseLegacyContextMenuStripSourceControlValue=false&quot;/&gt;&#13;&#10;&lt;/runtime&gt;&#13;&#10;</code></pre>Applications that target the .NET Framework 4.7.2 or later, and want to preserve the legacy behavior can opt in to the use of the legacy source control value by explicitly setting this AppContext switch to <code>true</code>.|
```

File: includes/migration-guide/retargeting/winforms/accessibility-improvements-windows-forms-controls-for-net-472.md [hunk 8]
```diff
@@ -3,7 +3,7 @@
...
...g><ul><li>The <xref:System.Windows.Forms.RadioButton> and <xref:System.Windows.Forms.CheckBox> controls with an empty <xref:System.Windows.Forms.ButtonBase.Text> property will now display a focus indicator when they receive focus.</li></ul><strong>Improved Property Grid Support</strong><ul><li>The <xref:System.Windows.Forms.PropertyGrid> control child elements now return a <code>true</code> for the  <xref:System.Windows.Automation.ValuePattern.IsReadOnlyProperty> property only when a PropertyGrid element is enabled.</li><li>The <xref:System.Windows.Forms.PropertyGrid> control child elements now return a <code>false</code> for the <xref:System.Windows.Automation.AutomationElement.IsEnabledProperty> property only when a PropertyGrid element can be changed by the user.</li></ul>For an overview of UI automation, see the [UI Automation Overview](~/docs/framework/ui-automation/ui-automation-overview.md).<strong>Improved keyboard navigation</strong><ul><li><xref:System.Windows.Forms.ToolStripButton> now allows focus when contained within a <xref:System.Windows.Forms.ToolStripPanel> that has the <xref:System.Windows.Forms.ToolStripPanel.TabStop> property set to <code>true</code></li></ul>|
```
Comment (NextTurn) [inline includes/migration-guide/retargeting/winforms/accessibility-improvements-windows-forms-controls-for-net-472.md]:
```suggestion
| Suggestion | <strong>How to opt in or out of these changes</strong>In order for the application to benefit from these changes, it must run on the .NET Framework 4.7.2 or later. The application can benefit from these changes in either of the following ways:<ul><li>It is recompiled to target the .NET Framework 4.7.2. These accessibility changes are enabled by default on Windows Forms applications that target the .NET Framework 4.7.2 or later.</li><li>It targets the .NET Framework 4.7.1 or earlier version and opts out of the legacy accessibility behaviors by adding the following [AppContext Switch](../../../../../docs/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element.md) to the <code>&lt;runtime&gt;</code> section of the app config file and setting it to <code>false</code>, as the following example shows.</li></ul><pre><code class="lang-xml">&lt;?xml version=&quot;1.0&quot; encoding=&quot;utf-8&quot;?&gt;&#13;&#10;&lt;configuration&gt;&#13;&#10;&lt;startup&gt;&#13;&#10;&lt;supportedRuntime version=&quot;v4.0&quot; sku=&quot;.NETFramework,Version=v4.7&quot;/&gt;&#13;&#10;&lt;/startup&gt;&#13;&#10;&lt;runtime&gt;&#13;&#10;&lt;!-- AppContextSwitchOverrides value attribute is in the form of &#39;key1=true/false;key2=true/false  --&gt;&#13;&#10;&lt;AppContextSwitchOverrides value=&quot;Switch.UseLegacyAccessibilityFeatures=false;Switch.UseLegacyAccessibilityFeatures.2=false&quot; /&gt;&#13;&#10;&lt;/runtime&gt;&#13;&#10;&lt;/configuration&gt;&#13;&#10;</code></pre>Note that to opt in to the accessibility features added in .NET Framework 4.7.2, you must also opt in to accessibility features of .NET Framework 4.7.1 as well. Applications that target the .NET Framework 4.7.2 or later and want to preserve the legacy accessibility behavior can opt in to the use of legacy accessibility features by explicitly setting this AppContext switch to <code>true</code>.<strong>Use of OS-defined colors in High Contrast themes</strong><ul><li>The drop down arrow of the <xref:System.Windows.Forms.ToolStripDropDownButton> now uses OS-defined colors in High Contrast theme.</li><li><xref:System.Windows.Forms.Button>, <xref:System.Windows.Forms.RadioButton> and <xref:System.Windows.Forms.CheckBox> controls with <xref:System.Windows.Forms.ButtonBase.FlatStyle> set to <xref:System.Windows.Forms.FlatStyle.Flat?displayProperty=nameWithType> or <xref:System.Windows.Forms.FlatStyle.Popup?displayProperty=nameWithType> now use OS-defined colors in High Contrast theme when selected. Previously, text and background colors were not contrasting and were hard to read.</li><li>Controls contained within a <xref:System.Windows.Forms.GroupBox> that has its <xref:System.Windows.Forms.Control.Enabled> property set to <code>false</code> will now use OS-defined colors in High Contrast theme.</li><li>The <xref:System.Windows.Forms.ToolStripButton>, <xref:System.Windows.Forms.ToolStripComboBox>, and <xref:System.Windows.Forms.ToolStripDropDownButton> controls have an increased luminosity contrast ratio in High Contrast Mode.</li><li><xref:System.Windows.Forms.DataGridViewLinkCell> will by default use OS-defined colors in High Contrast mode for the <xref:System.Windows.Forms.DataGridViewLinkCell.LinkColor?displayProperty=nameWithType> property.</li></ul>NOTE: Windows 10 has changed values for some high contrast system colors. Windows Forms Framework is based on the Win32 framework. For the best experience, run on the latest version of Windows and opt in to the latest OS changes by adding an app.manifest file in a test application and uncommenting the following code:<pre><code>&lt;!-- Windows 10 --&gt;&#13;&#10;&lt;supportedOS Id=&quot;{8e0f7a12-bfb3-4fe8-b9a5-48fd50a15a9a}&quot; /&gt;&#13;&#10;</code></pre><strong>Improved Narrator support</strong><ul><li>Narrator now announces the value of the <xref:System.Windows.Forms.ToolStripMenuItem.ShortcutKeys?displayProperty=nameWithType> property when announcing the text of a <xref:System.Windows.Forms.ToolStripMenuItem>.</li><li>Narrator now indicates when a <xref:System.Windows.Forms.ToolStripMenuItem> has its <xref:System.Windows.Forms.Control.Enabled> property set to <code>false</code>.</li><li>Narrator now gives feedback on the state of a check box when the <xref:System.Windows.Forms.ListView.CheckBoxes?displayProperty=nameWithType> property is set to <code>true</code>.</li><li>Narrator Scan Mode focus order is now consistent with the visual order of the controls on the ClickOnce download dialog window.</li></ul><strong>Improved DataGridView Accessibility support</strong><ul><li>Rows in a <xref:System.Windows.Forms.DataGridView> can now be sorted using the keyboard. Now a user can use the F3 key in order to sort by the current column.</li><li>When the <xref:System.Windows.Forms.DataGridView.SelectionMode?displayProperty=nameWithType> is set to <xref:System.Windows.Forms.DataGridViewSelectionMode.FullRowSelect?displayProperty=nameWithType>, the column header will change color to indicate the current column as the user tabs through the cells in the current row.</li><li>The <xref:System.Windows.Forms.DataGridViewCell.DataGridViewCellAccessibleObject.Parent?displayProperty=nameWithType> property now returns the correct parent control.</li></ul><strong>Improved Visual cues</strong><ul><li>The <xref:System.Windows.Forms.RadioButton> and <xref:System.Windows.Forms.CheckBox> controls with an empty <xref:System.Windows.Forms.ButtonBase.Text> property will now display a focus indicator when they receive focus.</li></ul><strong>Improved Property Grid Support</strong><ul><li>The <xref:System.Windows.Forms.PropertyGrid> control child elements now return a <code>true</code> for the  <xref:System.Windows.Automation.ValuePattern.IsReadOnlyProperty> property only when a PropertyGrid element is enabled.</li><li>The <xref:System.Windows.Forms.PropertyGrid> control child elements now return a <code>false</code> for the <xref:System.Windows.Automation.AutomationElement.IsEnabledProperty> property only when a PropertyGrid element can be changed by the user.</li></ul>For an overview of UI automation, see the [UI Automation Overview](../../../../../docs/framework/ui-automation/ui-automation-overview.md).<strong>Improved keyboard navigation</strong><ul><li><xref:System.Windows.Forms.ToolStripButton> now allows focus when contained within a <xref:System.Windows.Forms.ToolStripPanel> that has the <xref:System.Windows.Forms.ToolStripPanel.TabStop> property set to <code>true</code></li></ul>|
```

**Review by nxtn — COMMENTED** [review]
Since the destinations locate in the same repo, I found it's more common to use relative links that start with `../`.

And the `~` notation doesn't work on GitHub.

**Review by mairaw — APPROVED** [review]
Thanks for your changes @msschl and your feedback @NextTurn. I agree that the relative paths are preferred but they are not required, so I'll merge this as-is. For the files inside the includes folder though, the ~ are better because you don't know in which topic they'll be used, so in that case the use of ~ is better.

**Comment by NextTurn** [comment]
@mairaw Shall we add this special rule for `includes` folder to `styleguide/template.md#internal-links`?

**Comment by mairaw** [comment]
I think that would be good!

**Comment by NextTurn** [comment]
Well we have 4K links with `../includes/`.

https://github.com/dotnet/docs/search?q=..%2Fincludes%2F

**Comment by mairaw** [comment]
Using .. for the includes links in the content inside the docs folder is fine. My comment was regarding the links to our content from the files inside the includes like we had a few in this PR. For those files, we should use ~ for relative links. Makes sense?

**Comment by NextTurn** [comment]
Oh OK, I got it.

**Pull request merged**